README.md
utils
=====

[![Build Status](https://travis-ci.org/techjeffharris/utils.svg?branch=master)](https://travis-ci.org/techjeffharris/utils) 
[![Coverage Status](https://coveralls.io/repos/techjeffharris/utils/badge.svg?branch=master&service=github)](https://coveralls.io/github/techjeffharris/utils?branch=master)

handy utilties


CHANGES.md
utils CHANGES
=============

# v1.3.1
* [[`e5ee34d3f1`](https://github.com/techjeffharris/utils/commit/e5ee34d3f1)] - updated author field (Jeff Harris) 
* [[`33c0e22330`](https://github.com/techjeffharris/utils/commit/33c0e22330)] - trying to restore the explicit path to mocha (Jeff Harris) 
* [[`67510d29ba`](https://github.com/techjeffharris/utils/commit/67510d29ba)] - fixed the test directory in the after_script section (Jeff Harris) 
* [[`317755fa73`](https://github.com/techjeffharris/utils/commit/317755fa73)] - trying to get coveralls to work (Jeff Harris) 
* [[`e48e83af4f`](https://github.com/techjeffharris/utils/commit/e48e83af4f)] - fixed typo on line 8 (Jeff Harris) 
* [[`76f3d4c528`](https://github.com/techjeffharris/utils/commit/76f3d4c528)] - added tests and travis-ci support (Jeff Harris) 
* [[`730214563a`](https://github.com/techjeffharris/utils/commit/730214563a)] - 1.3.0 (Jeff Harris) 
* [[`c3a9d481b0`](https://github.com/techjeffharris/utils/commit/c3a9d481b0)] - adding exports for LatRegExp and LngRegExp (Jeff Harris) 
* [[`03fb066bbc`](https://github.com/techjeffharris/utils/commit/03fb066bbc)] - 1.2.0 (Jeff Harris) 
* [[`1064bf5ce4`](https://github.com/techjeffharris/utils/commit/1064bf5ce4)] - added LatRegExp and LngRegExp (Jeff Harris) 
* [[`5b5db763b1`](https://github.com/techjeffharris/utils/commit/5b5db763b1)] - 1.1.2 (Jeff Harris) 
* [[`52d5a20ba0`](https://github.com/techjeffharris/utils/commit/52d5a20ba0)] - Number.isNaN() is not broken, like Global isNaN() (Jeff Harris) 
* [[`dce6e920d2`](https://github.com/techjeffharris/utils/commit/dce6e920d2)] - told npm version minor not patch >_< (Jeff Harris) 
* [[`69d2d46d7a`](https://github.com/techjeffharris/utils/commit/69d2d46d7a)] - 1.2.0 (Jeff Harris) 
* [[`8a657ed231`](https://github.com/techjeffharris/utils/commit/8a657ed231)] - v1.1.1 - fixed GPSRegExp error (Jeff Harris) 
* [[`c79a9a3031`](https://github.com/techjeffharris/utils/commit/c79a9a3031)] - 1.1.0 (Jeff Harris) 
* [[`522339024c`](https://github.com/techjeffharris/utils/commit/522339024c)] - v1.1.0: added property GPSRegExp (Jeff Harris) 
* [[`0e16440444`](https://github.com/techjeffharris/utils/commit/0e16440444)] - v.1.0.3 - getType returns 'NaN' for NaN (Jeff Harris) 
* [[`52fbeb7e8c`](https://github.com/techjeffharris/utils/commit/52fbeb7e8c)] - v1.0.1 (Jeff Harris) 
* [[`1a28243767`](https://github.com/techjeffharris/utils/commit/1a28243767)] - **doc**: adding travis and coveralls badges to README.md (Jeff Harris) 


# v1.0.1
* Added GetType utility so I don't have to keep limping with typeof (credit to [Angus Croll](http://javascriptweblog.wordpress.com/2011/08/08/fixing-the-javascript-typeof-operator/) )
* Prototypal is unnecessary in nodejs enviroment.


test/gps-test.js

module.exports = function testGPS () {
  describe('#GPSRegExp', function describeGPSRegExp () {
    var GPSRegExp = require('../lib/utils').GPSRegExp;

    it('should strictly match latitude and longitude values that fall within the correct range', function testValidCoordinatePairs () {

      GPSRegExp.test('+90, -127.554334').should.be.true;
      GPSRegExp.test('45, 180').should.be.true;
      GPSRegExp.test('-90, -180').should.be.true;
      GPSRegExp.test('-90.000, -180.0000').should.be.true;
      GPSRegExp.test('+90, +180').should.be.true;
      GPSRegExp.test('47.1231231, 179.99999999').should.be.true;
      GPSRegExp.test('47.1231231,179.99999999').should.be.true;

      GPSRegExp.test('').should.be.false;
      GPSRegExp.test('-90., -180.').should.be.false;
      GPSRegExp.test('+90.1, -100.111').should.be.false;
      GPSRegExp.test('-91, 123.456').should.be.false;
      GPSRegExp.test('045, 180').should.be.false;

    });
  });

  describe('#LatRegExp', function describeLatRegExp () {
    var LatRegExp = require('../lib/utils').LatRegExp;

    it('should match valid latitude values', function testValidLatValues () {
      LatRegExp.test('90').should.be.true;
      LatRegExp.test('+90').should.be.true;
      LatRegExp.test('-90').should.be.true;
      LatRegExp.test('-90.000').should.be.true;
      LatRegExp.test('47.1231231').should.be.true;

      LatRegExp.test('').should.be.false;
      LatRegExp.test('-90.').should.be.false;
      LatRegExp.test('+90.1').should.be.false;
      LatRegExp.test('-91').should.be.false;
      LatRegExp.test('045').should.be.false;
    });
  });

  describe('#LngRegExp', function describeLngRegExp () {
    var LngRegExp = require('../lib/utils').LngRegExp;

    it('should match valid longitude values', function testValidLngValues () {
      LngRegExp.test('180').should.be.true;
      LngRegExp.test('+180').should.be.true;
      LngRegExp.test('-180').should.be.true;
      LngRegExp.test('-180.000').should.be.true;
      LngRegExp.test('179.99999999').should.be.true;

      LngRegExp.test('').should.be.false;
      LngRegExp.test('-180.').should.be.false;
      LngRegExp.test('+180.1').should.be.false;
      LngRegExp.test('-181').should.be.false;
      LngRegExp.test('045').should.be.false;
    });
  });
};


test/xor-test.js

module.exports = function testXOR () {
  describe('#XOR', function describeXOR () {

    var XOR = require('../lib/utils').XOR;

    it('should return true only if one value is truthy and the other is falsey', function () {
      XOR(true, false).should.be.true;
      XOR(false, true).should.be.true;
      XOR(true, true).should.be.false;
      XOR(false, false).should.be.false;
    });
  });
};


test/clone-test.js

module.exports = function testClone () {
  describe('#clone', function describeClone () {
    var clone = require('../lib/utils').clone;

    var parent;
    var cloned;
    var keys;

    it('should copy all of parents\' own properties into cloned', function () {
      parent = { 
        a: 1, 
        b: 'foo'
      };

      cloned = clone(parent);

      keys = Object.keys(cloned);

      keys.should.have.a.length(2);

      cloned.should.have.a.property('a', parent.a);
      cloned.should.have.a.property('b', parent.b);
      cloned.should.not.have.a.property
    });

    it('should return an object that is not equal to parent', function () {
      parent = {};
      cloned = clone(parent);

      (cloned === parent).should.be.false;
      (cloned == parent).should.be.false;

    });

    it('should return an empty object if parent is falsey', function () {
      parent = false; 
      cloned = clone(parent);

      (typeof cloned === 'object').should.be.true;

      keys = Object.keys(cloned);

      keys.should.have.length(0);

    });
  });
};


lib/utils.js

var BYTE  = 1;
var KB    = 1024 * BYTE;
var MB    = 1024 * KB;
var GB    = 1024 * MB;
var TB    = 1024 * GB;
var PB    = 1024 * TB;

var MILLISECOND = 1;
var SECOND      = 1000  * MILLISECOND;
var MINUTE      = 60    * SECOND;
var HOUR        = 60    * MINUTE;
var DAY         = 24    * HOUR;
var WEEK        = 7     * DAY;
var MONTH       = 30    * DAY;
var YEAR        = 365   * DAY;

// 
// "This one will strictly match latitude and longitude values that fall within the correct range:"
// ...
// Matches
// 
// +90.0, -127.554334 
// 45, 180
// -90, -180
// -90.000, -180.0000
// +90, +180
// 47.1231231, 179.99999999
// Doesn't Match
// 
// -90., -180.
// +90.1, -100.111
// -91, 123.456
// 045, 180
// 
// @see: http://stackoverflow.com/a/18690202/1690165
// 
var GPSRegExp = new RegExp(/^[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?),\s*[-+]?(180(\.0+)?|((1[0-7]\d)|([1-9]?\d))(\.\d+)?)$/)

// matches valid latitude strings
var LatRegExp = new RegExp(/^[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?)$/);

// matches valid longitude strings
var LngRegExp = new RegExp(/^[-+]?(180(\.0+)?|((1[0-7]\d)|([1-9]?\d))(\.\d+)?)$/);

// @see - http://killdream.github.io/blog/2011/10/understanding-javascript-oop/index.html

// Aliases for the rather verbose methods on ES5
var descriptor  = Object.getOwnPropertyDescriptor, 
  properties  = Object.getOwnPropertyNames, 
  define_prop = Object.defineProperty;

function clone(parent) {

  parent = parent || {};

  var cloned = {};
  
  properties(parent).forEach(function(key) {
    define_prop(cloned, key, descriptor(parent, key)) 
  });

  return cloned;
};

function extend(original, extensions) {       

  original = original || {};
  extensions = extensions || {};

  properties(extensions).forEach(function(key) {
    define_prop(original, key, descriptor(extensions, key)) 
  });

  return original;
};

function getType (obj) {
  return (Number.isNaN(obj)) ? 'NaN' : ({}).toString.call(obj).match(/\s([a-zA-Z]+)/)[1].toLowerCase()
};

// returns that to which an XOR expression would evaluate
function XOR (a,b) {
  return ( a || b ) && !( a && b );
};


module.exports = {

  clone: clone,

  // emitter: function () {
  //     var cloned 
  // }
  extend: extend,
  getType: getType,
  XOR: XOR,
  GPSRegExp: GPSRegExp,
  LatRegExp: LatRegExp,
  LngRegExp: LngRegExp,

  BYTE: BYTE,
  KB: KB,
  MB: MB,
  GB: GB,
  TB: TB,
  PB: PB,
  
  MILLISECOND: MILLISECOND,
  SECOND: SECOND,
  MINUTE: MINUTE,
  HOUR: HOUR,
  DAY: DAY,
  WEEK: WEEK,
  MONTH: MONTH,
  YEAR: YEAR
};
